### PR TITLE
feat: [#504] Route supports configure timeout

### LINKED ISSUE #504: ✨ [Feature] Route supports configure timeout
### Before feedback (在反馈之前)

- [X] There are no features that I want to submit in Issues (当前 Issues 中没有我将要提交的新功能)

### Describe Feature (描述功能)

1. A global configuration;
2. Add a new `WithContext` method for `ctx.Request()`, the new ctx is generated by `context.WithTimeout`, then the new method can be called in a middleware;

### PR DESCRIPTION
**Pull request by KlassnayaAfrodita**
https://github.com/goravel/goravel/issues/504

add withcontext

## 📑 Description

Closes https://github.com/goravel/goravel/issues/504

@coderabbitai summary

## ✅ Checks

- [ ] Added test cases for my code

### CODE REVIEW

**Review by hwbrzzl — COMMENTED** [review]
This new method is good, please add the global configuration logic.

**Comment by hwbrzzl** [comment]
Please merge your two PRs.

**Comment by KlassnayaAfrodita** [comment]
I merged two PCs. Please explain what "add the global configuration logic" means

**Comment by hwbrzzl** [comment]
image

**Comment by KlassnayaAfrodita** [comment]
I added a timeout to the config, how to use them in gin and fiber?

**Comment by KlassnayaAfrodita** [comment]
Do I understand correctly what needs to be added to goravel/gin/route.go in the function NewRoute(config config.Config, parameters map[string]any) (*Route, error) add 
// ...
    router.Use(gin.WithTimeout(config.Timeout))
// ...
?

**Comment by hwbrzzl** [comment]
CI failed.

**Comment by hwbrzzl** [comment]
gin doesn't have the `WithTimeout` method.

**Comment by KlassnayaAfrodita** [comment]
do I need to implement TimeoutMiddleware?

**Comment by hwbrzzl** [comment]
Yes, don't use gin-contrib/timeout, it has a lot of dependencies, we can implement it ourselves.

**Comment by KlassnayaAfrodita** [comment]
> Yes, don't use gin-contrib/timeout, it has a lot of dependencies, we can implement it ourselves.

where to place middleware timeout?

**Comment by hwbrzzl** [comment]
Follow: https://github.com/goravel/gin/blob/master/tls.go

**Comment by KlassnayaAfrodita** [comment]
I added timeout middleware following https://github.com/goravel/gin/blob/master/tls.go and rewrote the use of this middleware in NewRoute. Please check my changes and give feedback

File: timeoutMiddleware.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		select {
+		case <-done:
+			// The request completed before the timeout expired
+		case <-timeoutCtx.Done():
+			// After the timeout expires, return the status 504 Gateway Timeout
+			if timeoutCtx.Err() == context.DeadlineExceeded {
+				ctx.Response().Writer().WriteHeader(http.StatusGatewayTimeout)
+				_, _ = ctx.Response().Writer().Write([]byte("Request timed out"))
+			}
+		}
```
Comment (hwbrzzl) [inline timeoutMiddleware.go:38]: I doubt this's unnecessary, when we set `timeCtx` to `gin`, `gin` will handle this automatically. We can test it.

File: timeoutMiddleware.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"time"
+
+	contractshttp "github.com/goravel/framework/contracts/http"
+	"github.com/goravel/framework/contracts/config"
+)
+
+// TimeoutMiddleware creates middleware to set a timeout for a request
+func TimeoutMiddleware() contractshttp.Middleware {
+	return func(ctx contractshttp.Context) {
+		timeout := config.GetInt("http.http.drivers.gin.timeout", 3) * time.Second
```
Comment (hwbrzzl) [inline timeoutMiddleware.go:15]:
```suggestion
		timeout := config.GetInt("http.request_timeout", 3) * time.Second
```

File: timeoutMiddleware.go (hunk 1 above)
Comment (KlassnayaAfrodita) [reply timeoutMiddleware.go:38]: I'll trust your commit, where can I read about what you say?

**Comment by KlassnayaAfrodita** [comment]
what else should I do about this issue?

File: timeoutMiddleware.go (hunk 1 above)
Comment (KlassnayaAfrodita) [reply timeoutMiddleware.go:38]: I tested on localhost, gin does not cancel the request when WithTimeout expires. for example:
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

func main() {
	router := gin.Default()

	router.Use(func(c *gin.Context) {
		timeoutCtx, cancel := context.WithTimeout(c.Request.Context(), 3*time.Second)
		defer cancel()
		c.Request = c.Request.WithContext(timeoutCtx)
		c.Next()
	})

	router.GET("/long_request", func(c *gin.Context) {
		time.Sleep(6 * time.Second)
		c.String(http.StatusOK, "This should time out")
	})

	router.Run(":8080")
}

curl http://localhost:8080/long_request

the server continues processing the request after the context is canceled

Comment (hwbrzzl) [reply timeoutMiddleware.go:38]: Good catch 👍

File: timeoutMiddleware.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+		// Refreshing the request context with a timeout
+		ctx.WithContext(timeoutCtx)
+
+		// We start executing the request in a new goroutine
+		done := make(chan struct{})
+		go func() {
+			defer close(done)
+			ctx.Request().Next()
+		}()
```
Comment (hwbrzzl) [inline timeoutMiddleware.go:27]: Could you try the logic below? Not sure if the routine is required.

```suggestion
		ctx.Request().Next()
```

File: timeoutMiddleware.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+		select {
+		case <-done:
```
Comment (hwbrzzl) [inline timeoutMiddleware.go:30]: And modify this with above simultaneously.

```suggestion
		case <-ctx.Request().Origin().Context().Done():
```

File: timeoutMiddleware.go
Comment (hwbrzzl) [inline timeoutMiddleware.go:1]: The file name can be `timeout_middleware.go`, and change `tls.go` to `tls_middleware.go`.

File: timeoutMiddleware.go (hunk 4 above)
Comment (KlassnayaAfrodita) [reply timeoutMiddleware.go:30]: of course, so I need to remove the select block?

Comment (hwbrzzl) [reply timeoutMiddleware.go]: No, as you tested, only setting context is not enough, `select` is required. I mean changing `done` to `ctx.Request().Origin().Context().Done()`, test it, check if it's correct.

Comment (KlassnayaAfrodita) [reply timeoutMiddleware.go]: I tested it, it worked, I'll make changes today

File: timeout_middleware.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+		ctx.WithContext(timeoutCtx)
+
+		ctx.Request().Next()
+
+		select{
+		case <-ctx.Request().Origin().Context().Done():
+			
+			if timeoutCtx.Err() == context.DeadlineExceeded {
+				ctx.Response().Writer().WriteHeader(http.StatusGatewayTimeout)
+				_, _ = ctx.Response().Writer().Write([]byte("Request timed out"))
```
Comment (hwbrzzl) [inline timeout_middleware.go:28]:
```suggestion
				ctx.Request().AbortWithStatus(http.StatusGatewayTimeout)
```

File: route.go [hunk 6]
```diff
@@ -77,7 +81,7 @@ func (r *Route) Fallback(handler httpcontract.HandlerFunc) {
 }
 
 func (r *Route) GlobalMiddleware(middlewares ...httpcontract.Middleware) {
-	middlewares = append(middlewares, Cors(), Tls())
+	middlewares = append(middlewares, Cors(), Tls(), TImeoutMiddleware())
```
Comment (hwbrzzl) [inline route.go:84]:
```suggestion
	middlewares = append(middlewares, Cors(), Tls(), TimeoutMiddleware())
```

File: route.go [hunk 7]
```diff
@@ -30,6 +30,10 @@ func NewRoute(config config.Config, parameters map[string]any) (*Route, error) {
 	engine := gin.New()
 	engine.MaxMultipartMemory = int64(config.GetInt("http.drivers.gin.body_limit", 4096)) << 10
 	engine.Use(gin.Recovery()) // recovery middleware
+	// timeoutMiddleware
+	// timeout := TimeoutMiddleware(config.GetInt("http.request_timeout", 3) * time.Second)
+	// engine.Use(middlewaresToGinHandlers([]httpcontract.Middleware{timeout})...)
+
```
Comment (hwbrzzl) [inline route.go:36]: Remove it.

File: route.go
Comment (hwbrzzl) [inline route.go:1]: Please add test cases for the new logic in `route_test.go`.

**Comment by KlassnayaAfrodita** [comment]
please check my changes

File: route.go (hunk 6 above)
Comment (hwbrzzl) [reply route.go:81]: Please update this.

File: timeout_middleware.go (hunk 5 above)
Comment (hwbrzzl) [reply timeout_middleware.go:28]: Please update this.

**Review by hwbrzzl — COMMENTED** [review]
Okay, according to the change in this PR, I think it's time to merge https://github.com/goravel/framework/pull/675, please fix the CI.

After the PR above is merged, please execute `go get github.com/goravel/framework@master` in this PR, to make sure the CI of this PR passes.

**Comment by KlassnayaAfrodita** [comment]
https://github.com/goravel/gin/pull/97#discussion_r1807611415 https://github.com/goravel/gin/pull/97#discussion_r1807611511 already implemented

**Comment by KlassnayaAfrodita** [comment]
Please, check route_test.go. I think TimeoutMiddleware is not being called

**Comment by KlassnayaAfrodita** [comment]
with this version, everything works for me

**Comment by KlassnayaAfrodita** [comment]
I constantly get a null pointer to the request, that is, when ctx.Request() is called.Next() gives an error that the pointer is null

**Comment by KlassnayaAfrodita** [comment]
I constantly get a null pointer to the request, that is, when ctx.Request() is called.Next() gives an error that the pointer is null

**Comment by hwbrzzl** [comment]
Could you add me to this repo? https://github.com/KlassnayaAfrodita/gin
I optimized your code and want to push my code to your repo, so that you can check the optimization.

**Comment by KlassnayaAfrodita** [comment]
> hwbrzzl

yes, I invited you. patch-1 current branch

**Comment by hwbrzzl** [comment]
Please solve the conflict.

File: middleware_timeout_test.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+func TestTimeoutMiddleware(t *testing.T) {
+	mockConfig := mocksconfig.NewConfig(t)
+	mockConfig.EXPECT().GetBool("app.debug").Return(true).Once()
+	mockConfig.EXPECT().GetInt("http.drivers.gin.body_limit", 4096).Return(4096).Once()
+
+	route, err := NewRoute(mockConfig, nil)
+	require.NoError(t, err)
+
+	route.Middleware(Timeout(1*time.Second)).Get("/timeout", func(ctx contractshttp.Context) contractshttp.Response {
+		time.Sleep(1 * time.Second)
```
Comment (hwbrzzl) [inline middleware_timeout_test.go:24]: The test case failed in Windows: https://github.com/goravel/gin/actions/runs/11534959849/job/32113680328?pr=97

We may try to change this to 2 second:

```suggestion
		time.Sleep(2 * time.Second)
```

image

**Review by hwbrzzl — APPROVED** [review]
Great  PR, thanks!

**Comment by hwbrzzl** [comment]
We can optimize fiber next.

**Pull request merged**